### PR TITLE
Call static methods and build callbacks in MavenClassLoader

### PR DESCRIPTION
**Pull request by ivankelly**
To properly test some functionality we need to be able to call static
methods (e.g. for the creation of ByteBuf), and also to be able to
build callbacks (e.g. to call async methods).

This patch implements methods for doing this. For static methods, you
need to pass the className, the method name and a list of
arguments. It first tries to find the method using non-primitive
types, but falls back to trying primative types if possible.

Invokation looks like:
```
cl.callStaticMethod("org.apache.bookkeeper.client.BKException",
                    "create", [-101])
```

For callbacks you specify an interface and the a groovy closure.
It is assumed the closure will take as many arguments as the
callback. The invokation handler doesn't check the method name
called on the callback object, so it only works with callbacks
with a single method.

Invokation looks like:
```
def callback = classLoader.createCallback(
    "org.apache.bookkeeper.client.AsyncCallback\$AddCallback",
        { rc, _ledger, entryId, ctx ->
            // do something
        })
```

### CODE REVIEW

**Comment by ivankelly** [comment]
retest this please // to pull in integrationtests

File: tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java [hunk 1]
```diff
@@ -102,6 +106,47 @@ public static MavenClassLoader forBookKeeperVersion(String version) throws Excep
...
+
+    public Object callStaticMethod(String className, String methodName, ArrayList<?> args) throws Exception {
+        Class<?> klass = Class.forName(className, true, classloader);
+
+        try {
+            Class<?>[] paramTypes = args.stream().map((a)-> a.getClass()).toArray(Class[]::new);
+            return klass.getMethod(methodName, paramTypes).invoke(null, args.stream().toArray(Object[]::new));
+        } catch (NoSuchMethodException nsme) {
+            // maybe the params are primitives
+            Class<?>[] paramTypes = args.stream().map((a) -> {
```
Comment (eolivelli) [inline tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java]: I never used such kind of code for coercions. Do you this it is needed?
I will look into my code and send some pointer

Comment (ivankelly) [reply tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java]: It is. int.class is different to Integer.class, so without this, it won't find the method. I hit it when trying to find the static constructor for BKException. Groovy passed in an Integer as the argument. The method itself takes an int.

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

File: tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java (hunk 1 above)
Comment (eolivelli) [reply tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java]: Sure, I know, my point was about this way of doing coercions which I never seen and I wrote lot of code about this problems.
Any way this works and I think we can keep your proposal. I have no time today and this weekend and no access to laptop.

File: tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java [hunk 2]
```diff
@@ -102,6 +106,47 @@ public static MavenClassLoader forBookKeeperVersion(String version) throws Excep
...
+                    Class<?> k = a.getClass();
+                    try {
+                        Object type = k.getField("TYPE").get(null);
+                        if (type instanceof Class<?>) {
+                            return (Class<?>)type;
+                        } else {
+                            return k;
+                        }
+                    } catch (IllegalAccessException|NoSuchFieldException nsfe) {
```
Comment (eolivelli) [inline tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java]: nit: IllegalAccessException(space)|(space)NoSuchFieldException

File: tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java (hunk 1 above)
Comment (ivankelly) [reply tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java]: There's probably a much better way to do this, I just went for the quickest that works.

File: tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java (hunk 2 above)
Comment (ivankelly) [reply tests/integration-tests-utils/src/main/java/org/apache/bookkeeper/tests/MavenClassLoader.java]: There's nothing in the sun style guide about this, but in the tutorial there's no spaces. Personally I prefer to leave out the spaces, as it clusters all the types together visually.

https://docs.oracle.com/javase/tutorial/essential/exceptions/catch.html

**Comment by ivankelly** [comment]
retest this please // docker wasn't running on integration node

**Comment by ivankelly** [comment]
retest this please (have deleted individual retry job)

**Comment by ivankelly** [comment]
@sijie added test for the static method stuff. For the callback is more awkward, as it implies having groovy. the compat test that uses it will implicitly test it though

**Comment by sijie** [comment]
retest this please

**Comment by sijie** [comment]
all CI passed. integration test is still pending (due to too many integration tests are in the queue running). since this change is about maven class loader, it has nothing to do with integration tests, so ignore integration test CI to merge this pull request.

**Comment by sijie** [comment]
IGNORE CI

**Comment by ivankelly** [comment]
MavenClassloader is _only_ used in integration tests. Should be fine though, as nothing existing is modified.

**Pull request closed without merging**